File: backend/app/schemas/metrics.py
"""Pydantic schemas for metrics API endpoints

This module defines request/response models for the metrics feature
(Feature 002).
"""
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field, field_validator


# ============================================================================
# Time Series Query Schemas (FR-017)
# ============================================================================

class MetricsQueryParams(BaseModel):
    """Query parameters for time series endpoint"""
    metric_type: str = Field(..., description="Metric type to query")
    granularity: Literal['hourly', 'daily'] = Field(..., description="Data granularity")
    start_time: datetime = Field(..., description="Start of time range (inclusive)")
    end_time: datetime = Field(..., description="End of time range (inclusive)")

    @field_validator('start_time', 'end_time')
    @classmethod
    def ensure_timezone_aware(cls, v: datetime) -> datetime:
        """Ensure datetime is timezone-aware"""
        if v.tzinfo is None:
            raise ValueError("시간대 정보가 필요합니다 (ISO 8601 형식)")
        return v


class MetricDataPoint(BaseModel):
    """Single metric data point"""
    metric_type: str
    value: int
    granularity: str
    timestamp: datetime

    class Config:
        from_attributes = True


class MetricsTimeSeriesResponse(BaseModel):
    """Response for time series query"""
    metric_type: str
    granularity: str
    start_time: datetime
    end_time: datetime
    data_points: list[MetricDataPoint]
    total: int = Field(..., description="Total number of data points")


# ============================================================================
# Current Metrics Schemas (FR-007)
# ============================================================================

class CurrentMetricValue(BaseModel):
    """Current value for a single metric"""
    metric_type: str
    value: int | None = Field(..., description="Latest value, null if no data collected yet")
    display_name: str = Field(..., description="Korean display name")
    unit: str = Field(..., description="Unit of measurement")


class CurrentMetricsResponse(BaseModel):
    """Response for current metrics endpoint"""
    metrics: list[CurrentMetricValue]
    collected_at: datetime | None = Field(..., description="Timestamp of latest collection")


# ============================================================================
# Collection Status Schemas (FR-019)
# ============================================================================

class MetricFailureInfo(BaseModel):
    """Information about a failed collection"""
    metric_type: str
    attempted_at: datetime
    error_message: str


class MetricsCollectionStatusResponse(BaseModel):
    """Response for collection status endpoint"""
    last_collection_at: datetime | None
    next_collection_at: datetime
    recent_failures: list[MetricFailureInfo]
    failure_count_24h: int
    status: Literal['healthy', 'degraded']


# ============================================================================
# Summary Statistics Schemas
# ============================================================================

class MetricSummary(BaseModel):
    """Statistical summary for a metric over a time range"""
    metric_type: str
    granularity: str
    min: int | None
    max: int | None
    avg: int | None
    latest: int | None
    data_points: int


# ============================================================================
# Period Comparison Schemas (FR-022) - User Story 2
# ============================================================================

class PeriodInfo(BaseModel):
    """Information about a comparison period"""
    start_time: datetime
    end_time: datetime
    label: str = Field(..., description="Korean label (e.g., '이번 주', '지난 주')")
    data_points: list[MetricDataPoint]
    average: int | None


class MetricsComparisonResponse(BaseModel):
    """Response for period comparison endpoint"""
    metric_type: str
    granularity: str
    period1: PeriodInfo
    period2: PeriodInfo
    change_percentage: float | None = Field(
        ...,
        description="Percentage change from period2 to period1 ((p1-p2)/p2*100)"
    )
    change_direction: Literal['up', 'down', 'unchanged'] | None


# ============================================================================
# Export Schemas (FR-024, FR-025) - User Story 3
# ============================================================================

class ExportRequest(BaseModel):
    """Request for exporting metrics data"""
    metric_types: list[str] = Field(..., min_length=1, max_length=6)
    granularity: Literal['hourly', 'daily']
    start_time: datetime
    end_time: datetime
    format: Literal['csv', 'pdf']
    include_chart: bool = Field(
        default=False,
        description="Include chart in PDF export (ignored for CSV)"
    )

    @field_validator('start_time', 'end_time')
    @classmethod
    def ensure_timezone_aware(cls, v: datetime) -> datetime:
        """Ensure datetime is timezone-aware"""
        if v.tzinfo is None:
            raise ValueError("시간대 정보가 필요합니다 (ISO 8601 형식)")
        return v


class ExportResponse(BaseModel):
    """Response for export request"""
    export_id: str = Field(..., description="Unique export identifier")
    filename: str = Field(..., description="Generated filename")
    download_url: str = Field(..., description="URL to download the file")
    file_size_bytes: int
    expires_at: datetime = Field(..., description="Expiration time (1 hour from creation)")
    downsampled: bool = Field(
        default=False,
        description="Whether data was downsampled due to size limit"
    )

File: frontend/src/types/metrics.ts
/**
 * TypeScript interfaces for metrics feature (Feature 002)
 */

/**
 * Metric types tracked by the system
 */
export type MetricType =
  | 'active_users'
  | 'storage_bytes'
  | 'active_sessions'
  | 'conversation_count'
  | 'document_count'
  | 'tag_count';

/**
 * Granularity levels for metric snapshots
 */
export type MetricGranularity = 'hourly' | 'daily';

/**
 * Single metric data point (FR-017)
 */
export interface MetricSnapshot {
  metric_type: MetricType;
  value: number;
  granularity: MetricGranularity;
  timestamp: string; // ISO 8601 format
}

/**
 * Time range filter for metric queries (FR-017)
 */
export interface MetricTimeRange {
  metric_type: MetricType;
  granularity: MetricGranularity;
  start_time: string; // ISO 8601 format
  end_time: string; // ISO 8601 format
}

/**
 * Failed metric collection record (FR-019)
 */
export interface MetricCollectionFailure {
  metric_type: MetricType;
  attempted_at: string; // ISO 8601 format
  error_message: string;
}

/**
 * Metric configuration for UI display
 */
export interface MetricConfig {
  type: MetricType;
  displayName: string;
  unit: string;
  color: string; // Chart color
  formatValue: (value: number) => string; // Custom formatter (e.g., bytes to MB)
}

/**
 * Chart data structure for react-chartjs-2 (FR-017)
 */
export interface MetricChartData {
  labels: string[]; // Timestamps
  datasets: {
    label: string;
    data: number[];
    borderColor: string;
    backgroundColor: string;
    fill: boolean;
  }[];
}

/**
 * Period comparison data (FR-022)
 */
export interface PeriodComparisonData {
  metric_type: MetricType;
  current_period: {
    start: string;
    end: string;
    snapshots: MetricSnapshot[];
  };
  previous_period: {
    start: string;
    end: string;
    snapshots: MetricSnapshot[];
  };
  change_percentage: number; // Percentage change between periods
}

/**
 * Export format options (FR-024)
 */
export type ExportFormat = 'csv' | 'pdf';

/**
 * Export request parameters (FR-024, FR-025)
 */
export interface ExportRequest {
  metric_types: MetricType[];
  granularity: MetricGranularity;
  start_time: string; // ISO 8601 format
  end_time: string; // ISO 8601 format
  format: ExportFormat;
  include_chart?: boolean; // Only for PDF format
}

/**
 * API response wrapper for metric queries (FR-017)
 */
export interface MetricsResponse {
  data: MetricSnapshot[];
  total: number;
  granularity: MetricGranularity;
}

/**
 * API response for collection failures (FR-019)
 */
export interface FailuresResponse {
  failures: MetricCollectionFailure[];
  total: number;
}

/**
 * Downsampling options for chart performance (FR-018)
 */
export interface DownsampleOptions {
  enabled: boolean;
  threshold: number; // Max data points before downsampling
  algorithm: 'lttb'; // Largest-Triangle-Three-Buckets
}
